hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/TransactionMetadata.java

```diff
@@ -15,61 +15,97 @@
  */
 package com.hedera.node.app.spi.meta;
 
+import static java.util.Objects.requireNonNull;
+
 import com.hedera.node.app.spi.key.HederaKey;
 import com.hederahashgraph.api.proto.java.AccountID;
 import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
 import com.hederahashgraph.api.proto.java.TransactionBody;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
 import java.util.List;
+import java.util.Set;
 
 /**
  * Metadata collected when transactions are handled as part of "pre-handle". This happens with
  * multiple background threads. Any state read or computed as part of this pre-handle, including any
  * errors, are captured in the TransactionMetadata. This is then made available to the transaction
  * during the "handle" phase as part of the HandleContext.
+ *
+ * @param txnBody Transaction that is being pre-handled
+ * @param payer payer for the transaction
+ * @param status {@link ResponseCodeEnum} status of the transaction
+ * @param payerKey payer key required to sign the transaction. It is null if payer is missing
+ * @param requiredNonPayerKeys list of keys that are required to sign the transaction, in addition
+ *     to payer key
+ * @param readKeys the keys that were read during pre-handle
  */
-public interface TransactionMetadata {
-    /**
-     * Checks the failure by validating the status is not {@link ResponseCodeEnum OK}
-     *
-     * @return returns true if status is not OK
-     */
-    default boolean failed() {
-        return !status().equals(ResponseCodeEnum.OK);
+public record TransactionMetadata(
+        @Nullable TransactionBody txnBody,
+        @Nullable AccountID payer,
+        @NonNull ResponseCodeEnum status,
+        @Nullable HederaKey payerKey,
+        @NonNull List<HederaKey> requiredNonPayerKeys,
+        @Nullable Object handlerMetadata,
+        @NonNull List<ReadKeys> readKeys) {
+
+    public TransactionMetadata {
+        requireNonNull(status);
+        requireNonNull(requiredNonPayerKeys);
+        requireNonNull(readKeys);
     }
 
-    /**
-     * Returns the status of the transaction.
-     *
-     * @return the status of the transaction.
-     */
-    ResponseCodeEnum status();
+    public TransactionMetadata(
+            @NonNull final PrehandleHandlerContext context,
+            @NonNull final List<ReadKeys> readKeys) {
+        this(
+                context.getTxn(),
```

Comment:
null check for context

Reply:
`context.getTxn()` is a null check... kind of. 😏 

Do you prefer `requireNonNull(context).getTxn()`?